### PR TITLE
[Java] Support exchange ObjectRef between processes

### PR DESCRIPTION
**Pull request by ConeyLiu**
## Why are these changes needed?

This patch add supports exchanging ObjectRef between processes.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by kfstorm** [comment]
Thank you for submitting this PR. I believe this is what has been missing in #9634.

cc @stephanie-wang

**Review by stephanie-wang — APPROVED** [review]

File: java/runtime/src/main/java/io/ray/runtime/object/ObjectStore.java [hunk 1]
```diff
@@ -181,4 +186,19 @@ public abstract void delete(List<ObjectId> objectIds, boolean localOnly,
...
+   */
+  public abstract byte[] promoteAndGetOwnershipInfo(ObjectId objectId);
+
+  /**
+   * Register the given object ownership information.
+   * @param outer The outer ObjectId which contains the given ObjectId
+   * @param objectId The ID of the object needs to register
+   * @param ownership The serialized address of the owner
+   */
+  public abstract void registerObjectRef(ObjectId outer, ObjectId objectId, byte[] ownership);
```
Comment (kfstorm) [inline java/runtime/src/main/java/io/ray/runtime/object/ObjectStore.java:203]: Can you align it with the C++ declaration? e.g. the order of parameters, `outer` -> `outerObjectId`, `ownership` -> `ownerAddress`.

https://github.com/ray-project/ray/blob/427ba3df47f826f5114c81a98ff29ba1fc8efd13/src/ray/core_worker/core_worker.h#L444

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  @Test
+  public void testObjectTransfer() {
+    ObjectRef<String> objectRef = Ray.put("test");
+    List<ObjectRef<String>> data = new ArrayList<>();
+    data.add(objectRef);
+
+    ActorHandle<RemoteActor> handle = Ray.actor(RemoteActor::new).remote();
+    String result = handle.task(RemoteActor::get, data).remote().get();
+    Assert.assertEquals(result, "test");
+    handle.kill();
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:34]: You don't need to kill the actor manually. `Ray.shutdown()` is called in `BaseTest`.

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  public void testNestedObjectId() {
+    ObjectRef<String> inner = Ray.put("inner");
+    ObjectRef<ObjectRef<String>> outer = Ray.put(inner);
+    List<ObjectRef<ObjectRef<String>>> data = new ArrayList<>();
+    data.add(outer);
+
+    ActorHandle<RemoteActor> handle = Ray.actor(RemoteActor::new).remote();
+    String result = handle.task(RemoteActor::getNested, data).remote().get();
+    Assert.assertEquals(result, "inner");
+    handle.kill();
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:47]: Same as above.

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  public static class RemoteActor {
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:50]: No big deal. But normal tasks are simpler here.

File: src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc [hunk 5]
```diff
@@ -174,6 +174,35 @@ Java_io_ray_runtime_object_NativeObjectStore_nativeGetAllReferenceCounts(JNIEnv
...
+Java_io_ray_runtime_object_NativeObjectStore_nativePromoteAndGetOwnershipInfo(
+    JNIEnv *env, jclass, jbyteArray objectId) {
+  auto object_id = JavaByteArrayToId<ray::ObjectID>(env, objectId);
+  ray::CoreWorkerProcess::GetCoreWorker().PromoteObjectToPlasma(object_id);
+  ray::rpc::Address address;
+  ray::CoreWorkerProcess::GetCoreWorker().GetOwnershipInfo(object_id, &address);
+  auto address_str = address.SerializeAsString();
+  jbyteArray arr = env->NewByteArray(address_str.length());
+  env->SetByteArrayRegion(
+      arr, 0, address_str.length(), reinterpret_cast<const jbyte *>(address_str.c_str()));
```
Comment (kfstorm) [inline src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc:187]: Can you use `NativeStringToJavaByteArray` in `jni_utils.h` to simplify the code here?

File: src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc [hunk 6]
```diff
@@ -174,6 +174,35 @@ Java_io_ray_runtime_object_NativeObjectStore_nativeGetAllReferenceCounts(JNIEnv
...
+
+JNIEXPORT void JNICALL
+Java_io_ray_runtime_object_NativeObjectStore_nativeRegisterObjectRef(
+    JNIEnv *env, jclass, jbyteArray outer, jbyteArray objectId, jbyteArray ownerAddress) {
+  auto outer_objectId = JavaByteArrayToId<ray::ObjectID>(env, outer);
+  auto object_id = JavaByteArrayToId<ray::ObjectID>(env, objectId);
+  int len = env->GetArrayLength(ownerAddress);
+  std::string ownerAddressStr(len, 0);
+  env->GetByteArrayRegion(ownerAddress, 0, len,
+                          reinterpret_cast<jbyte *>(&ownerAddressStr.front()));
```
Comment (kfstorm) [inline src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc:199]: Can you use `JavaByteArrayToNativeString` in `jni_utils.h` to simplify the code here?

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 2 above)
Comment (kfstorm) [reply java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:34]: BTW, we also run tests under single-process mode. In this mode, the `kill` API is not available.

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class ObjectRefTransferTest extends BaseTest {
+
+  @BeforeClass
+  public void setUpClass() {
+    System.setProperty("ray.raylet.config.num_workers_per_process_java", "1");
```
Comment (raulchen) [inline java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:17]: This is not needed any more. We've change the default value to 1.

File: java/api/src/main/java/io/ray/api/id/ObjectId.java [hunk 8]
```diff
@@ -42,4 +44,13 @@ public int size() {
     return LENGTH;
   }
 
+  /**
+   * Generate a nil ObjectId.
+   */
+  private static ObjectId genNil() {
```
Comment (raulchen) [inline java/api/src/main/java/io/ray/api/id/ObjectId.java:50]: Nit: better avoid exposing nil ID in the Java API module. I think we can pass a null into `nativeRegisterObjectRef` instead, and convert null to nil ID in C++.

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 2 above)
Comment (raulchen) [inline java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java:34]: Don't need to kill the actor manually. The actor will be killed when the job finishes.

**Review by raulchen — COMMENTED** [review]
Thanks for this PR. Looks good overall. I just left a few small comments. Please fix them, then we can merge this PR.

File: java/runtime/src/main/java/io/ray/runtime/object/ObjectStore.java (hunk 1 above)
Comment (ConeyLiu) [reply java/runtime/src/main/java/io/ray/runtime/object/ObjectStore.java]: updated

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 2 above)
Comment (ConeyLiu) [reply java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java]: updated

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 3 above)
Comment (ConeyLiu) [reply java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java]: updated

File: src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc (hunk 5 above)
Comment (ConeyLiu) [reply src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc]: updated

File: src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc (hunk 6 above)
Comment (ConeyLiu) [reply src/ray/core_worker/lib/java/io_ray_runtime_object_NativeObjectStore.cc]: updated

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 7 above)
Comment (ConeyLiu) [reply java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java]: updated

File: java/api/src/main/java/io/ray/api/id/ObjectId.java (hunk 8 above)
Comment (ConeyLiu) [reply java/api/src/main/java/io/ray/api/id/ObjectId.java]: updated

File: java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java (hunk 2 above)
Comment (ConeyLiu) [reply java/test/src/main/java/io/ray/test/ObjectRefTransferTest.java]: updated

**Comment by ConeyLiu** [comment]
thanks @kfstorm @raulchen @stephanie-wang for the reviewing. Comments have addressed.

**Review by raulchen — APPROVED** [review]
LGTM. we'll merge it once CI passes.

**Review by kfstorm — APPROVED** [review]

**Comment by ConeyLiu** [comment]
thanks all.

**Pull request merged**